### PR TITLE
Add package evaluation fuzzer

### PR DESCRIPTION
**Pull request by instagibbs**
This fuzzer target caught the issue in https://github.com/bitcoin/bitcoin/pull/28251 within 5 minutes on master branch, and an additional issue which I've applied a preliminary patch to cover.

Fuzzer target does the following:

1) Picks mempool confgs, including max package size, count, mempool size, etc
2) Generates 1 to 26 transactions with arbitrary coins/fees, the first N-1 spending only confirmed outpoints
3) Nth transaction, if >1, sweeps all unconfirmed outpoints in mempool
4) If N==1, it may submit it through single-tx submission path, to allow for more interesting topologies
5) Otherwise submits through package submission interface
6) Repeat 1-5  a few hundred times per mempool instance

In other words, it ends up building chains of txns in the mempool using parents-and-children packages, which is currently the topology supported on master.

The test itself is a direct rip of tx_pool.cpp, with a number of assertions removed because they were failing for unknown reasons, likely due to the notification changes of single tx submission to package, which is used to track addition/removal of transactions in the test. I'll continue working on re-adding these assertions for further invariant testing.

### CODE REVIEW

**Comment by instagibbs** [comment]
cc @glozow @MarcoFalke

File: src/test/fuzz/package_eval.cpp
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:88]: Maybe worth adding `-bytespersigop` too?

**Comment by darosior** [comment]
Awesome, Concept ACK.

**Review by glozow — COMMENTED** [review]
concept ACK, really nice :rocket:

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:88]: to take advantage of this, probably would need to do some bare mulitisigs? wouldn't hurt to add I suppose

**Review by dergoegge — COMMENTED** [review]
Concept ACK

**Review by brunoerg — COMMENTED** [review]
Concept ACK

File: src/test/fuzz/package_eval.cpp
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:124]: In ddc6f41a1c72560b1d7701db921c0efb280c20b8: Is this `MakeMempool` the same from `tx_pool`? If so, perhaps we could move it to a util?

Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:124]: In ddc6f41a1c72560b1d7701db921c0efb280c20b8: It seems to have some functions from `tx_pool`. Couldn't we have them in a `util` instead of duplicating them?

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:123]: done

**Comment by instagibbs** [comment]
rebased on master, removed arg round-trip, un-marked WIP

**Comment by glozow** [comment]
Any interest in using `CheckPackageMempoolAcceptResult` from https://github.com/bitcoin/bitcoin/pull/26711/commits/baf475de599c82e7718ffefd7f0481fc88217853? It checks that a result contains the fields we expect based on whether it's supposed to be valid.

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [inline src/test/fuzz/package_eval.cpp:211]: making the tx >40kWu sometimes would have likely allowed the fuzzer to catch https://github.com/bitcoin/bitcoin/pull/28472

**Comment by instagibbs** [comment]
removed the second commit to not impede review of the various bugfix PRs individually

https://github.com/bitcoin/bitcoin/pull/28471
https://github.com/bitcoin/bitcoin/pull/28472

File: src/test/fuzz/package_eval.cpp
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:140]: I don't think this set is ever actually used.

File: src/test/fuzz/package_eval.cpp
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:295]: `Assert(true)` seems to not be useful.

File: src/test/fuzz/package_eval.cpp
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:115]: Seems a bit odd to me that the chain limits could go up to 50 but we only create packages of up to 26.

Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:115]: That's chain limits vs package limit checks, I think it's useful

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:140]: it is added to in `insert_tx` when used in `created_by_tx` arg. Lets each new package consider newly-entered outpoints in the mempool

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed debugging code

**Comment by instagibbs** [comment]
updated with cleanups

**Comment by instagibbs** [comment]
from discussions with others:

1) I have a PR built on https://github.com/bitcoin/bitcoin/pull/26711 which generalized from this fairly basic topology to general ancestor packages, including the last tx spending any subset of outpoints, and any package transactions potentially spending other package outpoints. I'd like to make sure it's getting relatively targeted coverage before switching to that, so deferring to after this PR.
2) We can add coverage where we mutate the wtxid of txns
3) Also should allow duplication of inputs

I plan on doing these in the follow-up on top of #26711

File: src/test/fuzz/package_eval.cpp
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:257]: In bf7436cde8ba7eac0912817f6f8a41828cb60bc2: In `ProcessNewPackage`, couldn't we just use `single_submit` for `/*test_accept=*/` directly, instead of creating `package_submit`?

File: src/test/fuzz/package_eval.cpp
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:155]: In bf7436cde8ba7eac0912817f6f8a41828cb60bc2: I think `amount_view` hasn't been used.

Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: deleted

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:256]: I find this ever so slightly more readable myself, opinions may vary

**Comment by instagibbs** [comment]
rebased on latest master on top of #28471 

Should not be crashing anymore :crossed_fingers: cc @dergoegge

File: src/test/fuzz/package_eval.cpp
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:179]:
```suggestion
                const auto num_in = last_tx ? package_outpoints.size() : fuzzed_data_provider.ConsumeIntegralInRange<int>(1, outpoints_rbf.size());
```

File: src/test/fuzz/package_eval.cpp
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:230]: Why is it calling `MockTime` again? I guess it's to trigger the `nLockTime`?

File: src/test/fuzz/package_eval.cpp
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:240]: In 2ecdc2183f0a1291730ae8728fa3d847b9cf80b5: Couldn't we prioritise more than 1 transaction?

File: src/test/fuzz/package_eval.cpp
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:276]: In `2ecdc2183f0a1291730ae8728fa3d847b9cf80b5`: Does it make sense to check whether the state is invalid if we previously assert that is valid?

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:230]: it's called for each package separately, sometimes, yes

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:240]: yes, I'm not sure if it will end up covering more interesting cases though

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:276]: sorry can you be a bit more explicit in what you're asking to do?

File: src/test/fuzz/package_eval.cpp
Comment (glozow) [reply src/test/fuzz/package_eval.cpp:253]: Also... wondering why allow `num_txs` to be less than 2 anyway? Isn't this test kind of the same as what's in the tx_pool fuzzer?

**Review by glozow — COMMENTED** [review]
A few quick questions, looks correct to me though some bits are a little confusing. I know future updates are planned so feel free to save for later if I'm scope creeping too much.

**Comment by glozow** [comment]
CI failure is a functional test so I'd assume unrelated

**Comment by dergoegge** [comment]
Made a coverage report: https://dergoegge.github.io/bitcoin-coverage/pr28450/fuzz.coverage/index.html

**Comment by glozow** [comment]
> Made a coverage report: https://dergoegge.github.io/bitcoin-coverage/pr28450/fuzz.coverage/index.html

Looks pretty good to me, noting the main missing bits:
- same-txid-different-witness
- policy script checks failing
- bad-txns-too-many-sigops
- being granted CPFP carve out and then failing

^which have already been marked as followups.

There's no package-not-child-with-unconfirmed-parents which is by design. Also looks like RBF rules are red, perhaps it'll take a while to hit them?

**Comment by dergoegge** [comment]
> Also looks like RBF rules are red, perhaps it'll take a while to hit them?

The seed corpus from the report was the result of 5000+ cpu hours, so I'd say it's likely just not able to hit the rbf paths.

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed as it's not very interesting a check

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: done

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:203]: done

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:244]: yes

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:310]: done

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: done, had to keep the outpoints_value check outside though since it needs all generated outpoints

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: > Also... wondering why allow num_txs to be less than 2 anyway?

1) it can get rejected by ProcessNewPackage
2) it can cause the existing mempool to be somewhat more interesting maybe

**Comment by instagibbs** [comment]
> Also looks like RBF rules are red, perhaps it'll take a while to hit them?

A couple of those cases would look difficult to hit without intention/design to do so. One I'm unsure how it hasn't been hit, something to look at later.

> I think achow is wondering if this is redundant with outpoints_rbf? Since new transactions are always created using outpoints_rbf inputs.

Can't find the actual place to reply(thanks github), removed outpoints_supply entirely as it's vestigial

----

Took all suggestions, pushed

File: src/test/fuzz/package_eval.cpp
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:28]: note: `g_outpoints_coinbase_init_immature` seems unused rn, I assume the plan is to maybe spend immature coinbases later?

File: src/test/fuzz/package_eval.cpp
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:199]: nit: `outpoints_rbf` contains unconfirmed, no?
```suggestion
            // Note that this test currently only spends package outputs in last transaction.
```

File: src/test/fuzz/package_eval.cpp
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:318]: nit: I think these `removed.erase()` calls might be unnecessary now that you don't use it to update `outpoints_rbf`.

**Review by glozow — COMMENTED** [review]
ACK 3707ed7c0d02d86ff86c159bbd746acfa52db145
Code and coverage look good. Will run overnight to sanity check that outpointsupdater hasn't broken anything.

Couple more ideas for post-26711 expansions:
- add nonexistent outpoint to hit missing inputs cases
- shuffle before submission

File: src/test/fuzz/package_eval.cpp
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:279]: `removed` doesn't seem to actually be used anywhere.

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: taken

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed `removed` entirely

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed `removed` entirely

**Comment by instagibbs** [comment]
Removed the `Final` subroutine to get a slight speedup since it seems unnecessary, leaving just the `tx_pool.check()` which should catch all issues.

**Comment by instagibbs** [comment]
Going to hold off on all subsequent nits to get this in to un-draft #26711 and start building on top

**Comment by instagibbs** [comment]
https://github.com/bitcoin/bitcoin/pull/28450/checks?check_run_id=17181111860 multiprocess build having an issue of some kind

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:96]: Sounds right, surprised I haven't hit this since it quite often chooses tiny mempool sizes too small for even a single transaction to enter.

Taken.

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed this already circa https://github.com/bitcoin/bitcoin/pull/28450/commits/70879e4f02e522ee82b563917ef6700efe4cf5dd

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:141]: was going to change this anyways, renaming to `mempool_outpoints`

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:161]: removed

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: ok fine, removed :)

**Comment by murchandamus** [comment]
ACK 262ab8ef7860d43cebc9d04721e3a075b4edf06e

Note that this is the first time I review something from the package relay project, so take that for whatever it’s worth

**Review by glozow — COMMENTED** [review]
reACK 262ab8ef7860d43cebc9d04721e3a075b4edf06e

**Review by dergoegge — APPROVED** [review]
tACK 262ab8ef7860d43cebc9d04721e3a075b4edf06e

Will generate and submit inputs to qa-assets once this is merged

File: src/test/fuzz/package_eval.cpp
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: Oh I was mistaken, AcceptPackage takes package sizes of 1(I was thinking of BIP331 probably!). This PR could have been a bit simpler I think.

**Pull request merged**
